java/src/jmri/util/SystemNameComparator.java

```diff
@@ -2,85 +2,47 @@
+        int p2len = Manager.getSystemPrefixLength(o2);
+
+        int comp = ac.compare(o1.substring(0, p1len), o2.substring(0, p2len));
+        if (comp != 0) return comp;
 
-        // if both strings are three or less characters long
-        if (o1.toString().length() <= 3 && o2.toString().length() <= 3) {
-            return o1.toString().compareTo(o2.toString());
-        } else  // if the first two characters of both strings match
-        if (!o1.toString().regionMatches(0, o2.toString(), 0, 2)) {
-            return o1.toString().compareTo(o2.toString());
-        } else {
-            if (true) {
-                AlphanumComparator ac = new AlphanumComparator();
-                return ac.compare(o1.toString(), o2.toString());
-            } else {    //TODO: dead-code strip this
-                // extract length of digits
-                char[] ch1 = o1.toString().substring(2).toCharArray();
-                char[] ch2 = o2.toString().substring(2).toCharArray();
-                int numDigit1 = 0;
-                int numDigit2 = 0;
-                for (int i = 0; i < ch1.length; i++) {
-                    if (Character.isDigit(ch1[i])) {
-                        numDigit1++;
-                    } else {
-                        break;
-                    }
-                }
-                for (int i = 0; i < ch2.length; i++) {
-                    if (Character.isDigit(ch2[i])) {
-                        numDigit2++;
-                    } else {
-                        break;
-                    }
-                }
-                if (numDigit1 == numDigit2) {
-                    try {
-                        int diff = Integer.parseInt(new String(ch1, 0, numDigit1))
-                                - Integer.parseInt(new String(ch2, 0, numDigit2));
-                        if (diff != 0) {
-                            return diff;
-                        }
-                        if (numDigit1 == ch1.length && numDigit2 == ch2.length) {
-                            return diff;
-                        } else {
-                            if (numDigit1 == ch1.length) {
-                                return -1;
-                            }
-                            // both have non-digit chars remaining
-                            return new String(ch1, numDigit1, ch1.length - numDigit1).compareTo(
-                                    new String(ch2, numDigit2, ch2.length - numDigit2));
-                        }
-                    } catch (NumberFormatException nfe) {
-                        return o1.toString().compareTo(o2.toString());
-                    } catch (IndexOutOfBoundsException ioob) {
-                        return o1.toString().compareTo(o2.toString());
-                    }
-                } else {
-                    return (numDigit1 - numDigit2);
-                }
-            }
-        }
+        char c1 = o1.charAt(p1len);
```

Comment:
These 5 lines seem unnecessarily complicated; why can’t it just: `return ac.compare(o1.substring(p1len), o2.substring(p2len));`? (What am I missing?)

Reply:
It's separating the sort of the type letter and the suffix, because that's what the comparison is documented to do. It's possible that the "chunking" in AlphanumComparator is invariant to that (I haven't thought it through systematically), but rather than working that out and documenting it, it seems simpler to me to have the code do what it says it should.